### PR TITLE
Automatically install Spring test dependencies into user project if required

### PR DESCRIPTION
**Pull request by EgorkaKulikov**
## Description

Supported an auto-installation of Spring or Spring Boot testing dependencies into user projects

Also:
- Added a setting to choose unit or integration testing
- "Framework will be installed" message is shown
- This setting is stored in State
- `LibraryMatcher` class refactored

### Manual tests

- Pure Spring project, no test dependency => it is installed with the same version as in project
- Pure Spring project, test dependency is present => no more dependencies installed
- Same two scenarios for Spring Boot projects

## Self-check list

- [x] I've set the proper **labels** for my PR (at least, for category and component).
- [x] PR **title** and **description** are clear and intelligible.
- [x] I've added enough **comments** to my code, particularly in hard-to-understand areas.
- [ ] The functionality I've repaired, changed or added is covered with **automated tests**.
- [ ] **Manual tests** have been provided optionally.
- [x] The **documentation** for the functionality I've been working on is up-to-date.

### CODE REVIEW

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/models/ExternalLibraryDescriptors.kt [hunk 1]
```diff
@@ -22,6 +22,13 @@ fun jUnit5ParametrizedTestsLibraryDescriptor(versionInProject: String?) =
 fun mockitoCoreLibraryDescriptor(versionInProject: String?) =
     ExternalLibraryDescriptor("org.mockito", "mockito-core", "3.5.0", null, versionInProject ?: "4.2.0")
 
+fun springBootTestLibraryDescriptor(versionInProject: String?) =
+    ExternalLibraryDescriptor("org.springframework.boot", "spring-boot-test", "2.4.0", null, versionInProject ?: "3.0.6")
+
+fun springTestLibraryDescriptor(versionInProject: String?) =
+    ExternalLibraryDescriptor("org.springframework", "spring-test", "2.4.0", null, versionInProject ?: "6.0.8")
```
Comment (IlyaMuravjov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/models/ExternalLibraryDescriptors.kt:29]: spring-test 2.4.0 doesn't really make sense here given the minimal version [mvnrepository is aware of is 2.5](https://mvnrepository.com/artifact/org.springframework/spring-test). I think `null` would be more self descriptive until we decide which minimal version we require.

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 2]
```diff
@@ -395,6 +401,13 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
             }
 
             if (model.projectType == ProjectType.Spring) {
+                row("Tests type:") {
+                    cell(sprintTestType)
+                    contextHelp(
+                        "Unit tests does not raise ApplicationContext <br>" +
+                                "while integration tests autowire beans."
```
Comment (IlyaMuravjov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:408]: I would work a bit more on the wording here, my initial suggestion is: "Unit tests do not  initialize `ApplicationContext` and not autowire beans, while integration tests do."

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 3]
```diff
@@ -861,11 +879,25 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
...
+        if (sprintTestType.item == SpringTestType.INTEGRATION_TESTS) {
+
+            if (SpringBoot.isInstalled) {
+                configureSpringTestDependency(SpringBoot)
+                return
+            }
+
+            if (SpringBeans.isInstalled) {
+                configureSpringTestDependency(SpringBeans)
+            }
```
Comment (IlyaMuravjov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:892]:
```suggestion
            configureSpringTestDependency(when {
                SpringBoot.isInstalled -> SpringBoot
                SpringBeans.isInstalled -> SpringBeans
                else -> else -> error("Unsupported DI framework type $framework")
            })
```

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 4]
```diff
@@ -884,11 +916,32 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
...
+        val frameworkTestVersionInProject = frameworkTestLibrary?.libraryName?.parseVersion()
+
+        if (frameworkTestVersionInProject == null || frameworkTestVersionInProject < frameworkVersionInProject) {
+            val libraryDescriptor = when (framework) {
+                SpringBoot ->  springBootTestLibraryDescriptor(frameworkVersionInProject)
+                SpringBeans -> springTestLibraryDescriptor(frameworkVersionInProject)
+                else -> error("Unsupported DI framework type $framework")
+            }
+
+            addDependency(model.testModule, libraryDescriptor)
```
Comment (IlyaMuravjov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:936]: `addDependency` returns `Promise<Void>`, here it's ignored, but we want the dependency to finish getting added at some point before `findPaths` gets called in `UtTestsDialogProcessor`.

**Review by IlyaMuravjov — CHANGES_REQUESTED** [review]

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt (hunk 2 above)
Comment (EgorkaKulikov) [reply utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt]: Seems `... and DO not autowire ...`, fixed

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt (hunk 3 above)
Comment (EgorkaKulikov) [reply utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt]: This fragment does not compile, but I created smth similar, thanks

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/models/ExternalLibraryDescriptors.kt (hunk 1 above)
Comment (EgorkaKulikov) [reply utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/models/ExternalLibraryDescriptors.kt]: I set version 2.5 as minimal. There are many bugs in Intellij dependency installation api, so we should better avoid testing what happens if we pass null as minimal version here ihmo.

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt (hunk 4 above)
Comment (EgorkaKulikov) [reply utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt]: Tried to fix using `blockingGet`, implemented it similar to the examples found in ide internals.

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 5]
```diff
@@ -884,11 +919,44 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
...
+            // Method `addDependency` returns Promise. Usually we need a dependency to be really installed
+            // only when tests are run by user. However, this Spring test dependency is required to generate tests,
+            // so we need to get the result of Promise as soon as possible. This is why we call `blockingGet` here.
+            try {
+                dependencyPromise.blockingGet(1000, TimeUnit.MILLISECONDS)
+            } catch (ignore: TimeoutException) {
+            } catch (e: ExecutionException) {
+                INTEGRATION_TESTS.frameworkInstalled = false
+                ExceptionUtil.rethrow(e)
+            }
```
Comment (IlyaMuravjov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:951]: If I am not mistaken this runs on the UI thread, so I advise moving blocking call elsewhere. For example, we may add `preCompilePromises: List<Promise<*>>` to `GenerateTestsModel`, save the promise in that list, and in `UtTestsDialogProcessor` replace `compile(...)`  with `preCompilePromises.all().thenAsync { compile(...) }`.

**Comment by EgorkaKulikov** [comment]
Still does not work correctly

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/generator/UtTestsDialogProcessor.kt [hunk 6]
```diff
@@ -181,243 +183,248 @@ object UtTestsDialogProcessor {
...
             .toTypedArray()
 
-        val promise = compile(project, filesToCompile, springConfigClass)
-        promise.onSuccess {
-            if (it.hasErrors() || it.isAborted)
-                return@onSuccess
+        model.preCompilePromises
+            .all()
+            .thenRun {
+                val compilationPromise = compile(project, filesToCompile, springConfigClass)
```
Comment (IlyaMuravjov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/generator/UtTestsDialogProcessor.kt:189]: Using fluent promise API can help avoid additional layer of nesting here
```suggestion
        val promise = model.preCompilePromises.all().thenAsync {
            compile(project, filesToCompile, springConfigClass)
        }
```

**Review by IlyaMuravjov — APPROVED** [review]

**Pull request merged**